build/provenance.go

```diff
@@ -0,0 +1,198 @@
+	controlapi "github.com/moby/buildkit/api/services/control"
+	"github.com/moby/buildkit/client"
+	provenancetypes "github.com/moby/buildkit/solver/llbsolver/provenance/types"
+	"github.com/opencontainers/go-digest"
+	ocispecs "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+)
+
+func setRecordProvenance(ctx context.Context, c *client.Client, sr *client.SolveResponse, ref string, pw progress.Writer) error {
+	mode := confutil.MetadataProvenance()
+	if mode == confutil.MetadataProvenanceModeNone {
+		return nil
+	}
+	pw = progress.ResetTime(pw)
+	return progress.Wrap("resolve build record provenance", pw.Write, func(l progress.SubLogger) error {
+		bo := backoff.NewExponentialBackOff()
+		bo.MaxElapsedTime = 5 * time.Second
+		return backoff.Retry(func() error {
+			res, err := fetchProvenance(ctx, c, ref, mode, l)
+			if err != nil {
+				return err
+			}
+			for k, v := range res {
+				sr.ExporterResponse[k] = v
+			}
+			return nil
+		}, bo)
+	})
+}
+
+func fetchProvenance(ctx context.Context, c *client.Client, ref string, mode confutil.MetadataProvenanceMode, l progress.SubLogger) (out map[string]string, err error) {
+	cl, err := c.ControlClient().ListenBuildHistory(ctx, &controlapi.BuildHistoryRequest{
+		Ref:       ref,
+		EarlyExit: true,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	var mu sync.Mutex
+	eg, ctx := errgroup.WithContext(ctx)
+	for {
+		ev, err := cl.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if ev.Record == nil {
+			continue
+		}
+		if ev.Record.Result != nil {
+			provenanceDgst := provenanceDigest(ev.Record.Result)
+			if provenanceDgst == nil {
+				continue
+			}
+			eg.Go(func() error {
+				dt, err := getBlob(ctx, c, *provenanceDgst, "", l)
+				if err != nil {
+					return errors.Wrapf(err, "failed to load provenance from build record")
+				}
+				prv, err := encodeProvenance(dt, mode)
+				if err != nil {
+					return err
+				}
+				mu.Lock()
+				if out == nil {
+					out = make(map[string]string)
+				}
+				out["buildx.build.provenance"] = prv
+				mu.Unlock()
+				return nil
+			})
+		} else if ev.Record.Results != nil {
+			for platform, res := range ev.Record.Results {
+				platform := platform
+				provenanceDgst := provenanceDigest(res)
+				if provenanceDgst == nil {
+					continue
+				}
+				eg.Go(func() error {
+					dt, err := getBlob(ctx, c, *provenanceDgst, platform, l)
+					if err != nil {
+						return errors.Wrapf(err, "failed to load provenance from build record")
+					}
+					prv, err := encodeProvenance(dt, mode)
+					if err != nil {
+						return err
+					}
+					mu.Lock()
+					if out == nil {
+						out = make(map[string]string)
+					}
+					out["buildx.build.provenance/"+platform] = prv
+					mu.Unlock()
+					return nil
+				})
+			}
+		}
+	}
+	return out, eg.Wait()
+}
+
+func provenanceDigest(res *controlapi.BuildResultInfo) *digest.Digest {
```

Comment:
`lookupProvenance(..)  *ocispecs.Descriptor` better I think.

Reply:
ReadBlob/ReaderAt takes a descriptor that can be used directly instead of rebuilding (a broken one) from a digest. Additionally I think the "Digest" part hurts readability as what this function really is doing is finding the provenance for a result. It does not return a digest for already resolved provenance descriptor for example. If you need a digest you can always just use `desc.Digest`